### ISSUE TITLE
Toy Model V0 - API

### ISSUE THREAD

**Issue by myyong** [alan-turing-institute/learningmachines#15]
**Description** Design the API for pulling data into toy model

**Criteria**
- [ ] Design docker-compose file
- [ ] Flask + SQL Alchemy for populating database
- [ ] API endpoints: Get number of patients
- [ ] Get single patient
- [ ] Get dataset for training/testing which consists of patient background, KFSS scores, binary 1/0 for relapse occurrence

**Comment by myyong** [comment]
What do we think of this:

### Directory structure
.
 * doc
 * infrastructure
   * toyModel0
      * client
         * html code
         * Dockerfile
      * model
         * Jupyter notebook to demo model @jsteyn this is for your notebook
         * access (this is how we would pipeline into model/autoprognosis)
           * API endpoints to send data to model
           * model code
         * Dockerfile
      * data
         * inputData
            * analysis @gordonyip427 this is for your CPATH notebook
            * access (this is how we would retrieve data for client toymodel, and for piping into model)
              * API endpoints to populate database, to retrieve data
            * Dockerfile
         * modelData 
            * access (this is how we would store data for model dataset )
              * API endpoints to store input data + outcome + prediction
            * Dockerfile
      * intepreter @MahedAb this is for your work
            * access
              * API endpoints to send data to intepreter and to receive output
            * Dockerfile
      * Docker-compose.yml This is how we could get the whole thing to work together.

**Comment by jack89roberts** [comment]
Looks like a sensible starting point to me!

Is retraining going to be part of the toy model? I think the model block above would just be taking one input/patient and giving a single prediction as output. Maybe there should be a dummy monitor block that counts how much new data there is, that triggers a retraining block, and then updates the model block? That implies being able to rebuild and redeploy the dockerfiles somehow.

Making a pipeline to train a model and then package the model into an API endpoint is something that tools like `MLFlow` should be good for. You can also track and visualise metrics about the performance of the trained models, but not sure whether it's possible to track/visualise the deployed model working on the new data.

**Comment by myyong** [comment]
>Is retraining going to be part of the toy model? I think the model block above would just be taking one input/patient and giving a single prediction as output. Maybe there should be a dummy monitor block that counts how much new data there is, that triggers a retraining block, and then updates the model block? That implies being able to rebuild and redeploy the dockerfiles somehow.

Agree. A *very* simple dummy monitor, just as a placeholder. That part is hard isn't it. I have experience with other bits, but not with that. @jsteyn 's scripting skills is going to come in very handy! 

> Making a pipeline to train a model and then package the model into an API endpoint is something that tools like MLFlow should be good for. You can also track and visualise metrics about the performance of the trained models, but not sure whether it's possible to track/visualise the deployed model working on the new data.

Thanks @jack89roberts that is really good to know! I can kind of see how the other bits work, but I hadn't a clue about that part.
